added metadata for flatten-array question

cloudleo opened the pull request:
first time adding metadata, would like people to review it and let me know if any problem exist

kytrinyx commented on flatten-array.md:
```diff
@@ -0,0 +1,19 @@
+The challenge is to write a function that accepts an arbitrarily-deep nested Array-like structure and returns a flattened structure with any null values removed without using the reduce method.
```
Not all languages have a reduce method. We'll need a way to phrase this so that it can be used for any languages.

Likewise, I think maybe we should say _list_ rather than _Array-like_. In Go, for example, we would probably use slices, not arrays for this.

kytrinyx commented on flatten-array.md:
```diff
@@ -0,0 +1,19 @@
+
+For Example
+
+flattener.flatten([0, 2, [[2, 3], 8, 100, null, [[null]]], -2]);
```
Here we would need to use some form of pseudo-code rather than something that can be taken for actual syntax.

kytrinyx commented on flatten-array.md:
```diff
@@ -0,0 +1,19 @@
+
+flattener.clear()
+
+returns []
```
I don't understand the purpose of `clear`. Can you elaborate?

kytrinyx commented on flatten-array.md:
```diff
@@ -0,0 +1,19 @@
+Tips
+
+* Define clear first before the function flatten
+* Think about closures and know what you have access to within a function
```
Not all languages have closures. If these tips are language or implementation-specific, they could be added to a `HINTS.md` file inside the exercise directory alongside the test suite and the reference solution. This would get included into the README for just that language track.

kytrinyx commented on flatten-array.yml:
```diff
@@ -0,0 +1,4 @@
+---
+blurb: "Write a program that will take a nested array and returns a single with all values but null "
```
Let's use _list_ instead of _array_. List feels like the more generic computer science-y term for this data structure.

Not all languages have nulls. Should we say _null or nil_ here?

kytrinyx commented on flatten-array.md:
```diff
@@ -0,0 +1,19 @@
+* Define clear first before the function flatten
+* Think about closures and know what you have access to within a function
+
+Have Fun, very popular interview question
```
Let's put this in source instead of "wolfram definition", but keep the source url pointing to the wolfram definition. The inspiration for this sounds like it was the interview question, rather than finding the definition on Wolfram.

kytrinyx commented on flatten-array.json:
```diff
@@ -0,0 +1,19 @@
+{
+   "cases": [
+      {
```
I wonder if there is a gentler build-up that we could use.

For example, what's the simplest possible thing? Perhaps `[]` -> `[]`. Then maybe `[[]]` -> `[]`.

kytrinyx commented on flatten-array.json:
```diff
@@ -0,0 +1,19 @@
+         "expected": [0, 2, 2, 3, 8, 100, -2]
+      },
+      {
+         "description": "flattens array with just integers present",
```
This seems like a simpler example than the previous one (no nulls vs has nulls). I think it should go before it.

kytrinyx commented on flatten-array.json:
```diff
@@ -0,0 +1,19 @@
+      {
+         "description": "flattens array with just integers present",
+         "input": [1,[2,3,4,5,6,7],8],
+         "expected": [1,2,3,4,5,6,7,8]
```
The smallest example of this would be `[1, [2], 3]` or, if you're worried about having a nested array of one element, maybe `[1, [2, 3], 4]`.

kytrinyx commented on flatten-array.json:
```diff
@@ -0,0 +1,19 @@
+      {
+         "description": "flattens array with null and integers present",
+         "input": [0, 2, [[2, 3], 8, 100, null, [[null]]], -2],
+         "expected": [0, 2, 2, 3, 8, 100, -2]
```
Here we have:

- an array
- a nested array
- a nested array inside the nested array
- nulls

I think we should build up the tests so that they are progressively more complex.

cloudleo commented:
I might have committed the same solution from earlier , please disregard that and close it.  I also updated the yml, md, and json.  Also simplified the example.js and the corresponding test with it

kytrinyx commented on flatten-array.json:
```diff
@@ -0,0 +1,29 @@
+         "expected": [1,2,3,4,5]
+      },
+      {
+         "description": "flattens array with just null",
```
This says there's null, but there are no nulls.

kytrinyx commented on flatten-array.json:
```diff
@@ -0,0 +1,29 @@
+         "expected":[1, 2, 3, 4, 5, 6, 7, 8]
+      },
+      {
+         "description": "flattens array with just null",
```
This has the same description as the previous one. In Ruby if you have several tests with the same name, then only the last one will be run.

Also, this doesn't have nulls in it.

kytrinyx commented on flatten-array.json:
```diff
@@ -0,0 +1,29 @@
+         "expected":[0, 2, 2, 3, 8, 100, 4, 50, -2]
+      },
+      {
+         "description": "flattens array with just null",
```
Same here. Same name, and no nulls.

There used to be nulls in the data, did that get lost?

cloudleo replied:
Sorry had problems last night trying to change the pull request from the console, decided to keep it more simple and get rid of null to be more general for all other language

kytrinyx replied:
I think that removing nulls while flattening is an interesting add-on. We just have to make sure to say null or nil in the problem description to make it language-unspecific.

cloudleo replied:
Got it! I just made changes to it and added null back in , thanks again @kytrinyx , if you see anything more else, please  don't hesitate to let me know, this is very fun!

kytrinyx commented on flatten-array.json:
```diff
@@ -0,0 +1,29 @@
+         "expected": [1,2,3,4,5]
+      },
+      {
+         "description": "flattens a deep nested list",
```
How about specifying "two levels" of nesting?

kytrinyx commented on flatten-array.json:
```diff
@@ -0,0 +1,29 @@
+         "expected":[1, 2, 3, 4, 5, 6, 7, 8]
+      },
+      {
+         "description": "flattens a deeper nested list",
```
Five levels of nesting.

kytrinyx commented on flatten-array.json:
```diff
@@ -0,0 +1,29 @@
+         "expected":[0, 2, 2, 3, 8, 100, 4, 50, -2]
+      },
+      {
+         "description": "flattens a deeper nested list part 2",
```
Four levels of nesting. It seems like four should come before five.

kytrinyx commented:
This looks good to me. Would you mind rebasing it onto master?

cloudleo commented:
Hey @kytrinyx  I am not sure how lol

kytrinyx commented:
Not a problem!

First figure out which remotes you have.

```
git remote -v
```

You'll probably have two lines, that both say `origin` and point to your fork (one for 'fetch', and one for 'push'). You'll need to add another origin, for this repository:

```
git remote add upstream https://github.com/exercism/x-common.git
```

Then pull down a local shadow copy of the upstream master branch:

```
git fetch upstream master
```

Then rebase what you have onto this local shadow copy:

```
git rebase upstream/master
```

You shouldn't have any conflicts.

That should be it. Holler if you run into trouble.

kytrinyx commented:
BTW: I mentioned this conversation in today's "Behind the Scenes" newsletter. http://tinyletter.com/exercism/letters/exercism-it-s-about-the-conversations

cloudleo commented:
@kytrinyx  So cool! I highly appreciate that, I rebase the code then pushed it, but its not showing up, not sure if I did it right or if you received should i do another pull request towards the master instead?

petertseng commented:
you pushed it to the "leo" branch, but this PR is from your repo's master branch, so you would to push to your repo's master branch before they are visible on this PR. That would be preferred approach rather than a new PR, since we can keep all the conversation in one place rather than spread out over two.

cloudleo commented:
Thanks @petertseng ! I was told it was good practice to do it in your own branch and I totally forgot to switch it back to master

kotp commented:
@cloudleo Any update on pushing master with this change (and squash?)?

cloudleo commented:
Its done, please check if everything is good @kotp and thanks for reviewing!!

kotp commented:
Thanks for all the hard work @cloudleo
